lib/VPackDeserializer/fixed-order.h

```diff
@@ -0,0 +1,178 @@
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Lars Maier
+///
+////////////////////////////////////////////////////////////////////////////////
+#ifndef VELOCYPACK_FIXED_ORDER_H
+#define VELOCYPACK_FIXED_ORDER_H
+#include "deserialize-with.h"
+#include "plan-executor.h"
+#include "utilities.h"
+
+namespace arangodb {
+namespace velocypack {
+namespace deserializer {
+
+/*
+ * Deserializes an array using a fixed order of deserializers. The result type
+ * is a tuple containing the result types of deserializers in order they appear
+ * in the template argument `Ds`.
+ */
+template <typename... Ds>
+struct fixed_order_deserializer {
+  using plan = fixed_order_deserializer<Ds...>;
+  using constructed_type = std::tuple<typename Ds::constructed_type...>;
+  using factory = utilities::identity_factory<constructed_type>;
+};
+
+template <typename... Ts>
+struct tuple_factory {
+  using constructed_type = std::tuple<Ts...>;
+
+  template <typename... S>
+  constructed_type operator()(S&&... s) {
+    return std::make_tuple(std::forward<S>(s)...);
+  }
+};
+
+template <typename... Ds>
+struct tuple_deserializer {
+  using constructed_type = std::tuple<typename Ds::constructed_type...>;
+  using plan = fixed_order_deserializer<Ds...>;
+  using factory = tuple_factory<typename Ds::constructed_type...>;
+};
+
+namespace executor {
+
+template <typename... Ds>
+struct plan_result_tuple<fixed_order_deserializer<Ds...>> {
+  using type = typename fixed_order_deserializer<Ds...>::constructed_type;
+};
+
+namespace detail {
+
+template <std::size_t I, typename T, typename E>
+struct fixed_order_deserializer_executor_visitor {
+  using value_type = std::optional<T>;
+
+  value_type& value_store;
+  E& error_store;
+  explicit fixed_order_deserializer_executor_visitor(value_type& value_store, E& error_store)
+      : value_store(value_store), error_store(error_store) {}
+
+  template <typename U = T>
+  bool operator()(U&& t) {
+    value_store.emplace(std::forward<U>(t));
+    return true;
+  }
+
+  bool operator()(E e) {
+    using namespace std::string_literals;
+    error_store = std::move(
+        e.wrap("in fixed order array at position "s + std::to_string(I)).trace(I));
+    return false;
+  }
+};
+
+}  // namespace detail
+
+template <typename... Ds, typename H>
+struct deserialize_plan_executor<fixed_order_deserializer<Ds...>, H> {
+  using value_type = typename fixed_order_deserializer<Ds...>::constructed_type;
+  using tuple_type = value_type;  // std::tuple<value_type>;
+  using result_type = result<tuple_type, deserialize_error>;
+
+  constexpr static auto expected_array_length = sizeof...(Ds);
```

Comment:
If the response to my following comment is "not needed, can be removed"
Then this value is of no need anymore, and can be removed as well.